deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py

```diff
@@ -0,0 +1,152 @@
+    an answer tuple and a background tuple, and calculates the degree to which the background tuple
+    `entails` the answer tuple.  In this layer, each input slot is represented as a dense encoding, so
+    to determine entailment we find the cosine similarity between these encoded slot representations,
+    i.e., the similarity between the first slot in each, then the second slot in each, etc.
+    This generates a set of similarity features equal to the number of slots in the tuples, which are
+    then fed to a shallow NN with output of size one.  The output of this NN is considered to be the
+    entailment score for the two tuples.
+
+    Inputs:
+        - tuple_1_input (the answer tuple), shape ``(batch size, num_slots, encoding_dim)``.  There also
+          needs to be a corresponding mask of shape (batch size, num_slots) (or None) that indicates whether
+          a given slot was all padding.
+
+        - tuple_2_input (the background_tuple), shape ``(batch size, num_slots, encoding_dim)``,
+          and again, there needs to be a corresponding mask of shape (batch size, num_slots) (or None)
+          that indicates whether a given slot was all padding.
+
+    Output:
+        - entailment score, shape ``(batch, 1)``
+
+    Parameters
+    ----------
+    - similarity_function_params: Dict[str, Any], default={}
+        These parameters get passed to a similarity function (see
+        :mod:`deep_qa.tensors.similarity_functions` for more info on what's acceptable).  The default
+        similarity function with no parameters is a simple dot product.
+
+    - num_hidden_layers : int, default=1
+        Number of hidden layers in the shallow NN.
+
+    - hidden_layer_width : int, default=4
+        The number of nodes in each of the NN hidden layers.
+
+    - initialization : string, default='glorot_uniform'
+        The initialization of the NN weights
+
+    - hidden_layer_activation : string, default='relu'
+        The activation of the NN hidden layers
+
+    - final_activation : string, default='sigmoid'
+        The activation of the NN output layer
+
+    """
+    def __init__(self, similarity_function: Dict[str, Any]=None, num_hidden_layers: int=1,
+                 hidden_layer_width: int=4, initialization: str='glorot_uniform',
+                 hidden_layer_activation: str='tanh', final_activation: str='sigmoid', **kwargs):
+        self.supports_masking = True
+        # Parameters for the shallow neural network
+        self.num_hidden_layers = num_hidden_layers
+        self.hidden_layer_width = hidden_layer_width
+        self.hidden_layer_init = initialization
+        self.hidden_layer_activation = hidden_layer_activation
+        self.final_activation = final_activation
+        self.hidden_layer_weights = []
+        self.score_layer = None
+        super(SlotSimilarityTupleMatcher, self).__init__(**kwargs)
+        sim_function_choice = get_choice_with_default(similarity_function,
```

Comment:
You need to copy what `MatrixAttention` does basically exactly.  In particular, you need a line like `if similarity_function is None: similarity_function = {}`.